File: scala/src/main/scala/com/interview/TwitterEtl.scala
package com.interview

import org.apache.log4j.{Level, Logger}
import org.apache.spark.SparkConf
import org.apache.spark.streaming.twitter.TwitterUtils
import org.apache.spark.streaming.{Seconds, StreamingContext}

object TwitterEtl {

  def main(args: Array[String]) {

    if (!Logger.getRootLogger.getAllAppenders.hasMoreElements) {
      Logger.getRootLogger.setLevel(Level.WARN)
    }

    //Create a SparkContext to initialize Spark
    val conf = new SparkConf()
      .setMaster("local[2]")
      .setAppName("Twitter ETL")

    val ssc = new StreamingContext(conf, Seconds(10))

    val stream = TwitterUtils.createStream(ssc, None)
    stream.flatMap(x => x.getText.split(" "))
        .map((_, 1))
        .reduceByKey((x, y) => x + y)
        .print()

    ssc.start()
    ssc.awaitTermination()
  }
}

File: java/src/java/com/interview/TwitterEtl.java
package com.interview;

import org.apache.log4j.Level;
import org.apache.log4j.Logger;
import org.apache.spark.SparkConf;
import org.apache.spark.streaming.Duration;
import org.apache.spark.streaming.api.java.JavaReceiverInputDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;
import org.apache.spark.streaming.twitter.TwitterUtils;
import scala.Tuple2;
import twitter4j.Status;

import java.util.Arrays;

public class TwitterEtl {

    public static void main(String [] args) throws Exception {
        if (!Logger.getRootLogger().getAllAppenders().hasMoreElements()) {
            Logger.getRootLogger().setLevel(Level.WARN);
        }

        SparkConf sparkConf = new SparkConf()
                .setMaster("local[2]")
                .setAppName("Twitter ETL");

        JavaStreamingContext ssc = new JavaStreamingContext(sparkConf, new Duration(10000));

        JavaReceiverInputDStream<Status> stream = TwitterUtils.createStream(ssc);
        stream.flatMap(x -> Arrays.asList(x.getText().split(" ")).iterator())
                .mapToPair(s -> new Tuple2<>(s, 1))
                .reduceByKey((x, y) -> x + y)
                .print();

        ssc.start();
        ssc.awaitTermination();
    }
}
